File: crates/kuben-api/src/routes/scope.rs
//! Resolve URL path segments into authorized scopes on the SQL model
//! (ADR-032). A project, environment or app is found by slug in each of the
//! caller's organizations in turn. One that does not exist, or exists only in
//! an organization the caller does not belong to, is `404`, not `403`, so its
//! existence does not leak.
//!
//! The scope chain names every node by its SQL id; a row imported from the
//! resource model keeps its Kubernetes UID as an alias, so role bindings and
//! tokens made on that UID keep applying until the importer rewrites them.
//! The resource projections only add live status: a row that was never
//! materialized has none yet.

use std::sync::Arc;

use kube::Client;
use kuben_core::{
    Error,
    authz::{ScopeChain, ScopeRef},
    ids::{ApplicationId, EnvironmentId, OrgId, ProjectId, TargetId},
};
use kuben_platform::{
    controller::resources::namespace_name,
    projection::{AppView, EnvironmentView, ProjectView},
};
use kuben_store::repo::{AppRecord, EnvironmentRecord, Project};

use crate::{authz::Authz, error::ApiError, state::ApiState};

fn not_found(kind: &str, name: &str) -> ApiError {
    ApiError(Error::NotFound(format!("{kind} `{name}`")))
}

/// The primary cluster, or `503` when none is configured.
pub fn cluster(state: &ApiState) -> Result<Client, ApiError> {
    state
        .cluster
        .as_ref()
        .map(kuben_platform::registry::ClusterRegistry::primary)
        .ok_or_else(|| ApiError(Error::Unavailable("no kubernetes cluster configured".into())))
}

/// Map a Kubernetes API error to a user-facing problem.
pub fn kube_error(err: kube::Error, name: &str) -> ApiError {
    match &err {
        kube::Error::Api(s) if s.code == 404 => not_found("object", name),
        kube::Error::Api(s) if s.code == 409 => ApiError(Error::Conflict(format!(
            "`{name}` already exists or was changed concurrently; retry"
        ))),
        kube::Error::Api(s) if s.code == 400 || s.code == 422 => {
            ApiError(Error::Validation(s.message.clone()))
        }
        _ => ApiError(Error::internal(err)),
    }
}

/// Kubernetes object name of an environment: `<project>-<env>`.
#[must_use]
pub fn environment_resource_name(project: &str, env: &str) -> String {
    format!("{project}-{env}")
}

/// Short environment name used in URLs and the UI.
#[must_use]
pub fn environment_short_name<'a>(project: &str, resource: &'a str) -> &'a str {
    resource
        .strip_prefix(project)
        .and_then(|r| r.strip_prefix('-'))
        .filter(|s| !s.is_empty())
        .unwrap_or(resource)
}

/// A live project of one of the caller's organizations.
#[derive(Debug)]
pub struct ProjectScope {
    pub org: OrgId,
    pub project: Project,
    /// Live status of its `Project` resource, once materialized.
    pub view: Option<Arc<ProjectView>>,
}

impl ProjectScope {
    #[must_use]
    pub fn chain(&self) -> ScopeChain {
        ScopeChain {
            aliases: self
                .project
                .legacy_uid
                .map(ScopeRef::Project)
                .into_iter()
                .collect(),
            ..ScopeChain::project(self.org, *self.project.id.as_uuid())
        }
    }

    #[must_use]
    pub fn slug(&self) -> &str {
        &self.project.slug
    }

    #[must_use]
    pub const fn id(&self) -> ProjectId {
        self.project.id
    }
}

/// A live environment of a project.
#[derive(Debug)]
pub struct EnvScope {
    pub project: ProjectScope,
    pub env: EnvironmentRecord,
    /// Live status of its `Environment` resource, once materialized.
    pub view: Option<Arc<EnvironmentView>>,
}

impl EnvScope {
    #[must_use]
    pub fn chain(&self) -> ScopeChain {
        let mut chain = self.project.chain();
        chain.environment = Some(*self.env.id.as_uuid());
        if let Some(uid) = self.env.legacy_uid {
            chain.aliases.push(ScopeRef::Environment(uid));
        }
        chain
    }

    #[must_use]
    pub fn short_name(&self) -> &str {
        &self.env.slug
    }

    /// Its `Environment` object's name, `<project>-<env>`.
    #[must_use]
    pub fn resource_name(&self) -> String {
        environment_resource_name(self.project.slug(), &self.env.slug)
    }

    /// Its placement's namespace.
    #[must_use]
    pub fn namespace(&self) -> String {
        self.env
            .namespace
            .clone()
            .unwrap_or_else(|| namespace_name(&self.resource_name()))
    }

    /// It, or its project, is being deleted.
    #[must_use]
    pub const fn deleting(&self) -> bool {
        self.env.deleting || self.project.project.deleting
    }

    #[must_use]
    pub const fn id(&self) -> EnvironmentId {
        self.env.id
    }
}

/// A live app: one application on an environment's placement.
#[derive(Debug)]
pub struct AppScope {
    pub env: EnvScope,
    pub app: AppRecord,
    /// Live status of its `App` resource, once materialized.
    pub view: Option<Arc<AppView>>,
}

impl AppScope {
    #[must_use]
    pub fn chain(&self) -> ScopeChain {
        let mut chain = self.env.chain();
        chain.app = Some(*self.app.target.as_uuid());
        if let Some(uid) = self.app.legacy_uid {
            chain.aliases.push(ScopeRef::App(uid));
        }
        chain
    }

    #[must_use]
    pub fn slug(&self) -> &str {
        &self.app.slug
    }

    #[must_use]
    pub fn namespace(&self) -> &str {
        &self.app.namespace
    }

    /// It, its environment or its project is being deleted.
    #[must_use]
    pub const fn deleting(&self) -> bool {
        self.app.deleting || self.env.deleting()
    }
}

/// The project `name` in the first of the caller's organizations that has one.
pub async fn project(state: &ApiState, authz: &Authz, name: &str) -> Result<ProjectScope, ApiError> {
    for org in authz.org_ids() {
        let mut tenant = state.store.tenant(org).await?;
        if let Some(project) = tenant.project(name).await? {
            let view = state
                .projections
                .project(&project.slug)
                .filter(|v| v.org.as_deref() == Some(org.to_string().as_str()));
            return Ok(ProjectScope { org, project, view });
        }
    }
    Err(not_found("project", name))
}

pub async fn environment(
    state: &ApiState,
    authz: &Authz,
    project: &str,
    env: &str,
) -> Result<EnvScope, ApiError> {
    let project = self::project(state, authz, project).await?;
    let mut tenant = state.store.tenant(project.org).await?;
    let env = tenant
        .environment(project.id(), env)
        .await?
        .ok_or_else(|| not_found("environment", env))?;
    let view = state
        .projections
        .environment(&environment_resource_name(project.slug(), &env.slug));
    Ok(EnvScope { project, env, view })
}

pub async fn app(
    state: &ApiState,
    authz: &Authz,
    project: &str,
    env: &str,
    app: &str,
) -> Result<AppScope, ApiError> {
    let env = environment(state, authz, project, env).await?;
    let mut tenant = state.store.tenant(env.project.org).await?;
    let record = tenant
        .app(env.id(), app)
        .await?
        .ok_or_else(|| not_found("app", app))?;
    let view = state.projections.app(&record.namespace, &record.slug);
    Ok(AppScope {
        env,
        app: record,
        view,
    })
}

/// The SQL ids of an app's target, for the deployment routes.
#[derive(Debug)]
pub struct TargetScope {
    pub org: OrgId,
    pub project: ProjectId,
    pub environment: EnvironmentId,
    pub application: ApplicationId,
    pub target: TargetId,
    chain: ScopeChain,
}

impl TargetScope {
    #[must_use]
    pub fn chain(&self) -> ScopeChain {
        self.chain.clone()
    }
}

/// The target of `app` in environment `env` of `project`.
pub async fn sql_target(
    state: &ApiState,
    authz: &Authz,
    project: &str,
    env: &str,
    app: &str,
) -> Result<TargetScope, ApiError> {
    let a = self::app(state, authz, project, env, app).await?;
    Ok(TargetScope {
        org: a.env.project.org,
        project: a.env.project.id(),
        environment: a.env.id(),
        application: a.app.application,
        target: a.app.target,
        chain: a.chain(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn short_names() {
        assert_eq!(environment_resource_name("shop", "prod"), "shop-prod");
        assert_eq!(environment_short_name("shop", "shop-prod"), "prod");
        assert_eq!(environment_short_name("shop", "legacy"), "legacy");
        assert_eq!(environment_short_name("shop", "shop-"), "shop-");
    }
}

File: go/hub/internal/api/scope.go
package api

import (
	"context"
	"strings"

	"github.com/Teamtem-dev/kuben/go/hub/internal/api/access"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/authz"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ids"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/kerr"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/projection"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/render"
	"github.com/Teamtem-dev/kuben/go/hub/internal/store"
)

// Scopes (routes/scope.rs): URL path segments resolved into authorized
// scopes on the SQL model (ADR-032). A project or environment is found by
// slug in each of the caller's organizations in turn; one that does not
// exist, or exists only in an organization the caller does not belong to,
// is 404, not 403, so its existence does not leak. The chain names every
// node by its SQL id; a row that came from the resource model keeps its
// Kubernetes UID as an alias. Live status from the projections is not
// ported yet (slice S1).

func scopeNotFound(kind, name string) error {
	return kerr.New(kerr.NotFound, "%s `%s`", kind, name)
}

// EnvironmentShortName is the environment name used in URLs and the UI:
// resource without its `<project>-` prefix, or resource itself when that
// would leave nothing.
func EnvironmentShortName(project, resource string) string {
	if rest, ok := strings.CutPrefix(resource, project); ok {
		if short, ok := strings.CutPrefix(rest, "-"); ok && short != "" {
			return short
		}
	}
	return resource
}

// projectScope is ProjectScope: a live project of one of the caller's
// organizations.
type projectScope struct {
	org     ids.OrgID
	project store.Project
	// view is the project's projection, when the cluster has it for this
	// organization.
	view opt.Val[projection.ProjectView]
}

func (p projectScope) chain() authz.ScopeChain {
	c := authz.ProjectChain(p.org, p.project.ID.UUID())
	if legacy, ok := p.project.LegacyUID.Get(); ok {
		c.Aliases = []authz.ScopeRef{{Level: authz.LevelProject, ID: legacy}}
	}
	return c
}

// findProject is scope::project: the project `name` in the first of the
// caller's organizations that has one.
func (s *Server) findProject(ctx context.Context, a access.Access, name string) (projectScope, error) {
	for _, org := range a.OrgIDs() {
		found, ok, err := s.projectNamed(ctx, org, name)
		if err != nil {
			return projectScope{}, err
		}
		if ok {
			return projectScope{org: org, project: found, view: s.projectView(org, found.Slug)}, nil
		}
	}
	return projectScope{}, scopeNotFound("project", name)
}

// projectView is the projection of project slug, unless it belongs to
// another organization than org.
func (s *Server) projectView(org ids.OrgID, slug string) opt.Val[projection.ProjectView] {
	v, ok := s.deps.Projections.Project(slug)
	if !ok || v.Org != opt.Some(org.String()) {
		return opt.None[projection.ProjectView]()
	}
	return opt.Some(*v)
}

func (s *Server) projectNamed(ctx context.Context, org ids.OrgID, name string) (store.Project, bool, error) {
	t, err := s.deps.Store.Tenant(ctx, org)
	if err != nil {
		return store.Project{}, false, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx)       //nolint:errcheck // read only
	return t.Project(ctx, name) //nolint:wrapcheck // a store error, answered as internal
}

// envScope is EnvScope: a live environment of a project.
type envScope struct {
	project projectScope
	env     store.EnvironmentRecord
	// view is the environment's projection, when the cluster has it.
	view opt.Val[projection.EnvironmentView]
}

// namespace is the environment's namespace: its placement's, else the one
// its controller creates.
func (e envScope) namespace() string {
	return e.env.Namespace.Or(render.NamespaceName(e.resourceName()))
}

// resourceName is the environment's Kubernetes object name.
func (e envScope) resourceName() string {
	return render.EnvironmentResourceName(e.project.project.Slug, e.env.Slug)
}

// deleting reports whether the environment, or its project, is being
// deleted (scope.rs EnvScope::deleting).
func (e envScope) deleting() bool {
	return e.env.Deleting || e.project.project.Deleting
}

func (e envScope) chain() authz.ScopeChain {
	c := e.project.chain()
	c.Environment = opt.Some(e.env.ID.UUID())
	if legacy, ok := e.env.LegacyUID.Get(); ok {
		c.Aliases = append(c.Aliases, authz.ScopeRef{Level: authz.LevelEnvironment, ID: legacy})
	}
	return c
}

// findEnvironment is scope::environment: environment env of the project
// `project` (found as findProject finds it).
func (s *Server) findEnvironment(ctx context.Context, a access.Access, project, env string) (envScope, error) {
	p, err := s.findProject(ctx, a, project)
	if err != nil {
		return envScope{}, err
	}
	t, err := s.deps.Store.Tenant(ctx, p.org)
	if err != nil {
		return envScope{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // read only
	record, found, err := t.Environment(ctx, p.project.ID, env)
	if err != nil {
		return envScope{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	if !found {
		return envScope{}, scopeNotFound("environment", env)
	}
	return envScope{
		project: p, env: record, view: s.environmentView(render.EnvironmentResourceName(p.project.Slug, record.Slug)),
	}, nil
}

// environmentView is the projection of the environment object resource.
func (s *Server) environmentView(resource string) opt.Val[projection.EnvironmentView] {
	v, ok := s.deps.Projections.Environment(resource)
	if !ok {
		return opt.None[projection.EnvironmentView]()
	}
	return opt.Some(*v)
}
